models/references.py

```diff
@@ -1,29 +1,46 @@
 import sqlalchemy as sa
 from sqlalchemy import orm
 
-from models.base import Base, AutoIDMixin
+from models.base import Base, AutoIDMixin, SmartSession
+from models.image import Image
+from models.provenance import Provenance
+from models.source_list import SourceList
+from models.psf import PSF
+from models.world_coordinates import WorldCoordinates
+from models.zero_point import ZeroPoint
 
 
-class ReferenceEntry(Base, AutoIDMixin):
+class Reference(Base, AutoIDMixin):
```

Comment:
When I was making comments elsewhere, I realized that with our provenance system we need more information here.

To see the problem, consider a subtraction.  A common use case will be a coadded reference, but a single new image.  So, the reference and the new image will *not* have the same provenance.  That means when loading the reference's source list (or wcs, or zp) you can't just say "give me the one that's the same as the one I'm using for the new image", because they will be different due to the reference and new image having different upstream provenances.  So, we don't know which provenance is the right one to ask for to get the reference image's source list (etc.).

The database normalization way to solve this is to first make zeropoint have an upstream of wcs (rather than soruce list).  Then, in this table, instead of image_id, you have zeropoint_id.  That will then uniqely define a (zeropoint, wcs, source list, image) that you use as a reference.  For convenience, I'd put in image, wcs, and source_list methods that crawl the join chain to get the right object out.

The scary way to solve this is to have four fields here: image_id, source_list_id, wcs_id, and zp_id.  Of course, then the database structure itself isn't enforcing consistency the way it does if we just link to a zeropoint.

Reply:
Discussion below.